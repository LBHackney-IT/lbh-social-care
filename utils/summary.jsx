const multiValue = (key, value, summaryInline) =>
  summaryInline ? (
    <span key={key}>{value} </span>
  ) : (
    <div key={key}>
      <span>{value}</span>
      <br />
    </div>
  );

const formatIsMulti = (formData, componentProps) => (
  <>
    {formData
      .map((multiData, index) => ({
        ...formatData(componentProps, {
          [componentProps.name]: multiData,
        }),
        key: `${componentProps.name}_${index}`,
      }))
      ?.map(({ key, value }) => (
        <div key={key}>{value}</div>
      ))}
  </>
);

const formatAddress = ({ address, postcode }, name, label) =>
  address && {
    key: name,
    title: label,
    value: (
      <>
        {address.split(', ').map((value) => (
          <div key={value}>
            <span>{value}</span>
            <br />
          </div>
        ))}
        <div>{postcode}</div>
      </>
    ),
  };

export const formatData = (componentProps, formData) => {
  const { component, options, name, label, isMulti, summaryInline } =
    componentProps;
  if (component === 'AddressLookup') {
    return formatAddress(formData[name], name, label);
  }
  if (component === 'Checkbox' && options) {
    return {
      key: name,
      title: label,
      value:
        typeof options[0] === 'string'
          ? formData[name]
          : formData[name]?.map((data) =>
              multiValue(
                data,
                options.find((o) => o.value === data)?.text,
                summaryInline
              )
            ),
    };
  }
  if (component === 'Radios' || component === 'Select') {
    const stepOptions =
      typeof options === 'function' ? options(formData) : options;
    return {
      key: name,
      title: label,
      value:
        !stepOptions || typeof stepOptions[0] === 'string'
          ? formData[name]
          : stepOptions.find((option) => option.value === formData[name])?.text,
    };
  }
  if (component === 'DateInput' || component === 'DatePicker') {
    return {
      key: name,
      title: label,
      value: new Date(formData[name]).toLocaleDateString('en-GB'),
    };
  }
  return {
    key: name,
    title: label,
    type: component,
    value: Array.isArray(formData[name])
      ? isMulti
        ? formatIsMulti(formData[name], componentProps)
        : formData[name].map((v) => multiValue(v, v))
      : typeof formData[name] === 'object'
      ? Object.entries(formData[name])
          .filter(([, value]) => Boolean(value))
          .map((entry) => multiValue(entry[0], entry[1], summaryInline))
      : typeof formData[name] === 'boolean'
      ? JSON.stringify(formData[name])
      : formData[name],
  };
};

export const getSectionObject = (formSteps, formData, value = false) =>
  formSteps.reduce((acc, { id, isMulti }) => {
    const stepIds = isMulti
      ? formData[id]?.reduce(
          (acc, _, key) => ({ ...acc, [`${id}/${key + 1}`]: false }),
          {}
        )
      : { [id]: value };
    return { ...acc, ...stepIds };
  }, {});
