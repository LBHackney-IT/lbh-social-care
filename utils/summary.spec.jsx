import { formatData, getSectionObject } from './summary';

describe('Summary utils', () => {
  describe('formatData', () => {
    it('should format "AddressLookup" properly', () => {
      const componentProps = {
        component: 'AddressLookup',
        name: 'main_address',
        label: 'main address',
      };
      const formData = {
        main_address: {
          address: 'first line, second line, third line',
          postcode: 'postcode',
          uprn: '123123',
        },
      };
      expect(formatData(componentProps, formData)).toMatchSnapshot();
    });

    it('should format "DateInput" properly', () => {
      const componentProps = {
        component: 'DateInput',
        name: 'first_date',
        label: 'Frst date',
      };
      const formData = {
        first_date: '2000-10-01',
      };
      expect(formatData(componentProps, formData)).toEqual({
        key: 'first_date',
        title: 'Frst date',
        value: '01/10/2000',
      });
    });

    it('should format "ObjectInput" properly', () => {
      const objectComponents = [
        {
          component: 'TextInput',
          name: 'first',
        },
        {
          component: 'TextInput',
          name: 'second',
        },
        {
          component: 'TextInput',
          name: 'third',
        },
      ];
      const componentProps = {
        component: 'ObjectInput',
        label: 'I am object component',
        name: 'oo',
        components: objectComponents,
      };
      const componentPropsInline = {
        component: 'ObjectInput',
        label: 'I am object component inline',
        name: 'oi',
        summaryInline: true,
        components: objectComponents,
      };
      const formData = {
        oo: { first: 'first', second: 'second', third: 'third' },
        oi: { first: 'first', second: 'second', third: 'third' },
      };
      expect(formatData(componentProps, formData)).toMatchSnapshot();
      expect(formatData(componentPropsInline, formData)).toMatchSnapshot();
    });

    it('should format "Checkbox" properly', () => {
      const componentProps = {
        component: 'Checkbox',
        label: 'I am checkbox component',
        name: 'checkbox',
        options: [
          { text: 'Foo', value: 'foo' },
          { text: 'Bar', value: 'bar' },
          { text: 'Foobar', value: 'foobar' },
          { text: 'Barfoo', value: 'barfoo' },
        ],
      };
      const formData = {
        checkbox: ['foo', 'bar', 'foobar'],
      };
      expect(formatData(componentProps, formData)).toMatchSnapshot();
    });

    it('should format "Select" properly', () => {
      const SelectProps = {
        component: 'Select',
        name: 'select',
        label: 'Select component',
        options: ['Yes', 'No'],
      };
      const SubSelectProps = {
        component: 'Select',
        name: 'subselect',
        label: 'Sub-Select component',
        options: ({ select }) =>
          select === 'Yes' && [
            { value: 'foo', text: 'I am Foo' },
            { value: 'bar', text: 'I am bar' },
          ],
      };
      const formData = {
        select: 'Yes',
        subselect: 'foo',
      };
      expect(formatData(SelectProps, formData)).toEqual({
        key: 'select',
        title: 'Select component',
        value: 'Yes',
      });
      expect(formatData(SubSelectProps, formData)).toEqual({
        key: 'subselect',
        title: 'Sub-Select component',
        value: 'I am Foo',
      });
    });

    it('should format "TextInput isMulti" properly', () => {
      const componentProps = {
        component: 'TextInput',
        name: 'text',
        label: 'Some_text',
        isMulti: true,
      };
      const formData = {
        text: ['i', 'm', 'multi', 'text'],
      };
      expect(formatData(componentProps, formData)).toMatchSnapshot();
    });
  });

  describe('getSectionObject', () => {
    it('should work properly', () => {
      const formSteps = [{ id: 'foo' }, { id: 'bar', isMulti: true }];
      const formData = { foo: '', bar: [1, 2, 3] };
      expect(getSectionObject(formSteps, formData)).toEqual({
        foo: false,
        'bar/1': false,
        'bar/2': false,
        'bar/3': false,
      });
    });
  });
});
