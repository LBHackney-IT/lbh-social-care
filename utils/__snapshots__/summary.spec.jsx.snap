// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Summary utils formatData should format "AddressLookup" properly 1`] = `
Object {
  "key": "main_address",
  "title": "main address",
  "value": <React.Fragment>
    <div>
      <span>
        first line
      </span>
      <br />
    </div>
    <div>
      <span>
        second line
      </span>
      <br />
    </div>
    <div>
      <span>
        third line
      </span>
      <br />
    </div>
    <div>
      postcode
    </div>
  </React.Fragment>,
}
`;

exports[`Summary utils formatData should format "Checkbox" properly 1`] = `
Object {
  "key": "checkbox",
  "title": "I am checkbox component",
  "value": Array [
    <div>
      <span>
        Foo
      </span>
      <br />
    </div>,
    <div>
      <span>
        Bar
      </span>
      <br />
    </div>,
    <div>
      <span>
        Foobar
      </span>
      <br />
    </div>,
  ],
}
`;

exports[`Summary utils formatData should format "ObjectInput" properly 1`] = `
Object {
  "key": "oo",
  "title": "I am object component",
  "type": "ObjectInput",
  "value": Array [
    <div>
      <span>
        first
      </span>
      <br />
    </div>,
    <div>
      <span>
        second
      </span>
      <br />
    </div>,
    <div>
      <span>
        third
      </span>
      <br />
    </div>,
  ],
}
`;

exports[`Summary utils formatData should format "ObjectInput" properly 2`] = `
Object {
  "key": "oi",
  "title": "I am object component inline",
  "type": "ObjectInput",
  "value": Array [
    <span>
      first
       
    </span>,
    <span>
      second
       
    </span>,
    <span>
      third
       
    </span>,
  ],
}
`;

exports[`Summary utils formatData should format "TextInput isMulti" properly 1`] = `
Object {
  "key": "text",
  "title": "Some_text",
  "type": "TextInput",
  "value": <React.Fragment>
    <div>
      i
    </div>
    <div>
      m
    </div>
    <div>
      multi
    </div>
    <div>
      text
    </div>
  </React.Fragment>,
}
`;
